{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set working directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/bartosz/theta_methyl\n"
     ]
    }
   ],
   "source": [
    "%cd ~/theta_methyl/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download FASTQ file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-03-20 22:11:01--  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR617/001/SRR6176401/SRR6176401.fastq.gz\n",
      "           => 'SRR6176401.fastq.gz'\n",
      "Resolving ftp.sra.ebi.ac.uk... 193.62.193.138\n",
      "Connecting to ftp.sra.ebi.ac.uk|193.62.193.138|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /vol1/fastq/SRR617/001/SRR6176401 ... done.\n",
      "==> SIZE SRR6176401.fastq.gz ... 1081731478\n",
      "==> PASV ... done.    ==> RETR SRR6176401.fastq.gz ... done.\n",
      "Length: 1081731478 (1.0G) (unauthoritative)\n",
      "\n",
      "SRR6176401.fastq.gz 100%[===================>]   1.01G  2.28MB/s    in 7m 13s  \n",
      "\n",
      "2023-03-20 22:18:15 (2.38 MB/s) - 'SRR6176401.fastq.gz' saved [1081731478]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR617/001/SRR6176401/SRR6176401.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR6176401.1 1/1\r\n",
      "TTTTATATTTTTATTTTTTATAAATTATAATTTATTTTTATTGTATTTATGGTATTGGATTAATAATATTAGTGTAGCTAAATTAGTTAAAAGATTAATG\r\n",
      "+\r\n",
      "BBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIFIIIIIIIIFFFIIIIIIIIIFFIIIIIIFIIFF07BFIIIFFFFFFFFFFFFFFFB\r\n"
     ]
    }
   ],
   "source": [
    "!head -4 <(gzcat SRR6176401.fastq.gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# conda environments:\r\n",
      "#\r\n",
      "base                  *  /Users/bartosz/anaconda3\r\n",
      "MachineLearningCourse     /Users/bartosz/anaconda3/envs/MachineLearningCourse\r\n",
      "filip                    /Users/bartosz/anaconda3/envs/filip\r\n",
      "methyl_theta             /Users/bartosz/anaconda3/envs/methyl_theta\r\n",
      "mothur                   /Users/bartosz/anaconda3/envs/mothur\r\n",
      "ngs2                     /Users/bartosz/anaconda3/envs/ngs2\r\n",
      "qiime2-2020.8            /Users/bartosz/anaconda3/envs/qiime2-2020.8\r\n",
      "r4-base                  /Users/bartosz/anaconda3/envs/r4-base\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!conda info --envs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fastp: an ultra-fast all-in-one FASTQ preprocessor\r\n",
      "version 0.12.4\r\n",
      "usage: fastp --in1=string [options] ... \r\n",
      "options:\r\n",
      "  -i, --in1                            read1 input file name (string)\r\n",
      "  -o, --out1                           read1 output file name (string [=])\r\n",
      "  -I, --in2                            read2 input file name (string [=])\r\n",
      "  -O, --out2                           read2 output file name (string [=])\r\n",
      "  -6, --phred64                        indicates the input is using phred64 scoring (it'll be converted to phred33, so the output will still be phred33)\r\n",
      "  -z, --compression                    compression level for gzip output (1 ~ 9). 1 is fastest, 9 is smallest, default is 2. (int [=2])\r\n",
      "  -A, --disable_adapter_trimming       adapter trimming is enabled by default. If this option is specified, adapter trimming is disabled\r\n",
      "  -a, --adapter_sequence               the adapter for read1. For SE data, if not specified, the adapter will be auto-detected. For PE data, this is used if R1/R2 are found not overlapped. (string [=auto])\r\n",
      "      --adapter_sequence_r2            the adapter for read2 (PE data only). This is used if R1/R2 are found not overlapped. If not specified, it will be the same as <adapter_sequence> (string [=])\r\n",
      "  -f, --trim_front1                    trimming how many bases in front for read1, default is 0 (int [=0])\r\n",
      "  -t, --trim_tail1                     trimming how many bases in tail for read1, default is 0 (int [=0])\r\n",
      "  -F, --trim_front2                    trimming how many bases in front for read2. If it's not specified, it will follow read1's settings (int [=0])\r\n",
      "  -T, --trim_tail2                     trimming how many bases in tail for read2. If it's not specified, it will follow read1's settings (int [=0])\r\n",
      "  -g, --trim_poly_g                    force polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data\r\n",
      "      --poly_g_min_len                 the minimum length to detect polyG in the read tail. 10 by default. (int [=10])\r\n",
      "  -G, --disable_trim_poly_g            disable polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data\r\n",
      "  -5, --cut_by_quality5                enable per read cutting by quality in front (5'), default is disabled (WARNING: this will interfere deduplication for both PE/SE data)\r\n",
      "  -3, --cut_by_quality3                enable per read cutting by quality in tail (3'), default is disabled (WARNING: this will interfere deduplication for SE data)\r\n",
      "  -W, --cut_window_size                the size of the sliding window for sliding window trimming, default is 4 (int [=4])\r\n",
      "  -M, --cut_mean_quality               the bases in the sliding window with mean quality below cutting_quality will be cut, default is Q20 (int [=20])\r\n",
      "  -Q, --disable_quality_filtering      quality filtering is enabled by default. If this option is specified, quality filtering is disabled\r\n",
      "  -q, --qualified_quality_phred        the quality value that a base is qualified. Default 15 means phred quality >=Q15 is qualified. (int [=15])\r\n",
      "  -u, --unqualified_percent_limit      how many percents of bases are allowed to be unqualified (0~100). Default 40 means 40% (int [=40])\r\n",
      "  -n, --n_base_limit                   if one read's number of N base is >n_base_limit, then this read/pair is discarded. Default is 5 (int [=5])\r\n",
      "  -L, --disable_length_filtering       length filtering is enabled by default. If this option is specified, length filtering is disabled\r\n",
      "  -l, --length_required                reads shorter than length_required will be discarded, default is 15. (int [=15])\r\n",
      "  -c, --correction                     enable base correction in overlapped regions (only for PE data), default is disabled\r\n",
      "  -U, --umi                            enable unique molecular identifer (UMI) preprocessing\r\n",
      "      --umi_loc                        specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none (string [=])\r\n",
      "      --umi_len                        if the UMI is in read1/read2, its length should be provided (int [=0])\r\n",
      "      --umi_prefix                     if specified, an underline will be used to connect prefix and UMI (i.e. prefix=UMI, UMI=AATTCG, final=UMI_AATTCG). No prefix by default (string [=])\r\n",
      "  -p, --overrepresentation_analysis    enable overrepresented sequence analysis.\r\n",
      "  -P, --overrepresentation_sampling    one in (--overrepresentation_sampling) reads will be computed for overrepresentation analysis (1~10000), smaller is slower, default is 20. (int [=20])\r\n",
      "  -j, --json                           the json format report file name (string [=fastp.json])\r\n",
      "  -h, --html                           the html format report file name (string [=fastp.html])\r\n",
      "  -R, --report_title                   should be quoted with ' or \", default is \"fastp report\" (string [=fastp report])\r\n",
      "  -w, --thread                         worker thread number, default is 3 (int [=3])\r\n",
      "  -s, --split                          split output by limiting total split file number with this option (2~999), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (int [=0])\r\n",
      "  -S, --split_by_lines                 split output by limiting lines of each file with this option(>=1000), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (long [=0])\r\n",
      "  -d, --split_prefix_digits            the digits for the sequential number padding (1~10), default is 4, so the filename will be padded as 0001.xxx, 0 to disable padding (int [=4])\r\n",
      "  -?, --help                           print this message\r\n"
     ]
    }
   ],
   "source": [
    "!fastp "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!open fastp.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download reference genome (let's download only one chromosome to make the analysis faster)\n",
    "###### https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-56/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-03-21 20:53:34--  https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-56/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz\n",
      "Resolving ftp.ensemblgenomes.ebi.ac.uk... 193.62.193.141\n",
      "Connecting to ftp.ensemblgenomes.ebi.ac.uk|193.62.193.141|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 9211107 (8.8M) [application/x-gzip]\n",
      "Saving to: 'Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz'\n",
      "\n",
      "Arabidopsis_thalian 100%[===================>]   8.78M  1.24MB/s    in 7.2s    \n",
      "\n",
      "2023-03-21 20:53:41 (1.23 MB/s) - 'Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz' saved [9211107/9211107]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-56/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">1 dna:chromosome chromosome:TAIR10:1:1:30427671:1 REF\r\n",
      "CCCTAAACCCTAAACCCTAAACCCTAAACCTCTGAATCCTTAATCCCTAAATCCCTAAAT\r\n",
      "CTTTAAATCCTACATCCATGAATCCCTAAATACCTAATTCCCTAAACCCGAAACCGGTTT\r\n",
      "CTCTGGTTGAAAATCATTGTGTATATAATGATAATTTTATCGTTTTTATGTAATTGCTTA\r\n",
      "TTGTTGTGTGTAGATTTTTTAAAAATATCATTTGAGGTCAATACAAATCCTATTTCTTGT\r\n",
      "GGTTTTCTTTCCTTCACTTAGCTATGGATGGTTTATCTTCATTTGTTATATTGGATACAA\r\n",
      "GCTTTGCTACGATCTACATTTGGGAATGTGAGTCTCTTATTGTAACCTTAGGGTTGGTTT\r\n",
      "ATCTCAAGAATCTTATTAATTGTTTGGACTGTTTATGTTTGGACATTTATTGTCATTCTT\r\n",
      "ACTCCTTTGTGGAAATGTTTGTTCTATCAATTTATCTTTTGTGGGAAAATTATTTAGTTG\r\n",
      "TAGGGATGAAGTCTTTCTTCGTTGTTGTTACGCTTGTCATCTCATCTCTCAATGATATGG\r\n"
     ]
    }
   ],
   "source": [
    "!head <(gzcat Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Imagine this file contains the whole reference genome. How to replace the header of FASTA file by presenting only:\n",
    "##### >1\n",
    "##### ...\n",
    "##### >2\n",
    "##### ...\n",
    "##### etc.? Replace the header by short chromosome name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/bartosz/theta_methyl\r\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "zcat: can't stat: genome/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz (genome/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz.Z): No such file or directory\r\n"
     ]
    }
   ],
   "source": [
    "!zcat genome/Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz | awk '/^>/{print \">\" ++i; next}{print}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mv Arabidopsis_thaliana.TAIR10.dna.chromosome.1.fa.gz genome/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing bisulfite genomes out into a single MFA (multi FastA) file\n",
      "\n",
      "Bisulfite Genome Indexer version v0.24.0 (last modified: 19 May 2022)\n",
      "\n",
      "A directory called /Users/bartosz/theta_methyl/genome/Bisulfite_Genome/ already exists. Already existing converted sequences and/or already existing Bowtie 2, HISAT2 or Minimap2) indices will be overwritten!\n",
      "\n",
      "\n",
      "Step I - Prepare genome folders - completed\n",
      "\n",
      "\n",
      "\n",
      "Total number of conversions performed:\n",
      "C->T:\t5435374\n",
      "G->A:\t5421151\n",
      "\n",
      "Step II - Genome bisulfite conversions - completed\n",
      "\n",
      "\n",
      "Bismark Genome Preparation - Step III: Launching the Bowtie 2 indexer\n",
      "Please be aware that this process can - depending on genome size - take several hours!\n",
      "Settings:\n",
      "  Output files: \"BS_CT.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genome_mfa.CT_conversion.fa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:01\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "Settings:\n",
      "  Output files: \"BS_GA.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genome_mfa.GA_conversion.fa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 7565828\n",
      "Using parameters --bmax 5674371 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 5674371 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  Time reading reference sizes: 00:00:01\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  V-Sorting samples time: 00:00:02\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Time to join reference sequences: 00:00:01\n",
      "bmax according to bmaxDivN setting: 7565828\n",
      "Using parameters --bmax 5674371 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 5674371 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 3.02633e+07 (target: 5674370)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 30263312 for bucket 1\n",
      "  (Using difference cover)\n",
      "  V-Sorting samples time: 00:00:01\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:01\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 3.02633e+07 (target: 5674370)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 30263312 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:28\n",
      "Returning block of 30263313 for bucket 1\n",
      "  Sorting block time: 00:00:29\n",
      "Returning block of 30263313 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 9709674\n",
      "fchr[G]: 9709674\n",
      "fchr[T]: 15130825\n",
      "fchr[$]: 30263312\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 14284280 bytes to primary EBWT file: BS_CT.1.bt2\n",
      "Wrote 7565836 bytes to secondary EBWT file: BS_CT.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 30263312\n",
      "    bwtLen: 30263313\n",
      "    sz: 7565828\n",
      "    bwtSz: 7565829\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 1891458\n",
      "    offsSz: 7565832\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 157622\n",
      "    numLines: 157622\n",
      "    ebwtTotLen: 10087808\n",
      "    ebwtTotSz: 10087808\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:36\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:01\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 7565828\n",
      "Using parameters --bmax 5674371 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 5674371 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:01\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:01\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 15130825\n",
      "fchr[G]: 20566199\n",
      "fchr[T]: 20566199\n",
      "fchr[$]: 30263312\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 14284280 bytes to primary EBWT file: BS_GA.1.bt2\n",
      "Wrote 7565836 bytes to secondary EBWT file: BS_GA.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 30263312\n",
      "    bwtLen: 30263313\n",
      "    sz: 7565828\n",
      "    bwtSz: 7565829\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 1891458\n",
      "    offsSz: 7565832\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 157622\n",
      "    numLines: 157622\n",
      "    ebwtTotLen: 10087808\n",
      "    ebwtTotSz: 10087808\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:38\n",
      "Reading reference sizes\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 3.02633e+07 (target: 5674370)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 30263312 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Time reading reference sizes: 00:00:01\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Time to join reference sequences: 00:00:01\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 7565828\n",
      "Using parameters --bmax 5674371 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 5674371 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:01\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:01\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 3.02633e+07 (target: 5674370)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 30263312 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:21\n",
      "Returning block of 30263313 for bucket 1\n",
      "  Sorting block time: 00:00:22\n",
      "Returning block of 30263313 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 9709674\n",
      "fchr[G]: 9709674\n",
      "fchr[T]: 15130825\n",
      "fchr[$]: 30263312\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 14284280 bytes to primary EBWT file: BS_CT.rev.1.bt2\n",
      "Wrote 7565836 bytes to secondary EBWT file: BS_CT.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 30263312\n",
      "    bwtLen: 30263313\n",
      "    sz: 7565828\n",
      "    bwtSz: 7565829\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 1891458\n",
      "    offsSz: 7565832\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 157622\n",
      "    numLines: 157622\n",
      "    ebwtTotLen: 10087808\n",
      "    ebwtTotSz: 10087808\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:30\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 15130825\n",
      "fchr[G]: 20566199\n",
      "fchr[T]: 20566199\n",
      "fchr[$]: 30263312\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 14284280 bytes to primary EBWT file: BS_GA.rev.1.bt2\n",
      "Wrote 7565836 bytes to secondary EBWT file: BS_GA.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 30263312\n",
      "    bwtLen: 30263313\n",
      "    sz: 7565828\n",
      "    bwtSz: 7565829\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 1891458\n",
      "    offsSz: 7565832\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 157622\n",
      "    numLines: 157622\n",
      "    ebwtTotLen: 10087808\n",
      "    ebwtTotSz: 10087808\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:31\n",
      "\n",
      "=========================================\n",
      "\n",
      "Parallel genome indexing complete. Enjoy!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!bismark_genome_preparation --bowtie2 genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detecting adapter for single end input...\n",
      "Detected adapter: TTATATTATGAAAGTTTTGAGAAGTA\n",
      "\n",
      "Read1 before filtering:\n",
      "total reads: 18885258\n",
      "total bases: 1888525800\n",
      "Q20 bases: 1832907690(97.0549%)\n",
      "Q30 bases: 1741003315(92.1885%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 18804298\n",
      "total bases: 1768245951\n",
      "Q20 bases: 1721381346(97.3497%)\n",
      "Q30 bases: 1640652789(92.7842%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 18804298\n",
      "reads failed due to low quality: 22823\n",
      "reads failed due to too many N: 6683\n",
      "reads failed due to too short: 51454\n",
      "reads with adapter trimmed: 936285\n",
      "bases trimmed due to adapters: 23013760\n",
      "\n",
      "JSON report: fastp.json\n",
      "HTML report: fastp.html\n",
      "\n",
      "fastp --in1 SRR6176401.fastq.gz --out1 SRR6176401.filtered.fastq.gz -t 5 \n",
      "fastp v0.12.4, time used: 188 seconds\n"
     ]
    }
   ],
   "source": [
    "!fastp --in1 SRR6176401.fastq.gz --out1 SRR6176401.filtered.fastq.gz -t 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "!open fastp.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bowtie 2 seems to be working fine (tested command 'bowtie2 --version' [2.4.2])\n",
      "Output format is BAM (default)\n",
      "Alignments will be written out in BAM format. Samtools found here: '/Users/bartosz/anaconda3/bin/samtools'\n",
      "Reference genome folder provided is genome/\t(absolute path is '/Users/bartosz/theta_methyl/genome/)'\n",
      "FastQ format assumed (by default)\n",
      "\n",
      "Input files to be analysed (in current folder '/Users/bartosz/theta_methyl'):\n",
      "SRR6176401.filtered.fastq.gz\n",
      "Library is assumed to be strand-specific (directional), alignments to strands complementary to the original top or bottom strands will be ignored (i.e. not performed!)\n",
      "Setting parallelization to single-threaded (default)\n",
      "\n",
      "Summary of all aligner options:\t-q --score-min L,0,-0.2 --ignore-quals\n",
      "Current working directory is: /Users/bartosz/theta_methyl\n",
      "\n",
      "Now reading in and storing sequence information of the genome specified in: /Users/bartosz/theta_methyl/genome/\n",
      "\n",
      "chr 1 (30427671 bp)\n",
      "\n",
      "Single-core mode: setting pid to 1\n",
      "\n",
      "Single-end alignments will be performed\n",
      "=======================================\n",
      "\n",
      "Input file is in FastQ format\n",
      "Writing a C -> T converted version of the input file SRR6176401.filtered.fastq.gz to SRR6176401.filtered.fastq.gz_C_to_T.fastq\n",
      "\n",
      "Created C -> T converted version of the FastQ file SRR6176401.filtered.fastq.gz (18811118 sequences in total)\n",
      "\n",
      "Input file is SRR6176401.filtered.fastq.gz_C_to_T.fastq (FastQ)\n",
      "\n",
      "Now running 2 instances of Bowtie 2 against the bisulfite genome of /Users/bartosz/theta_methyl/genome/ with the specified options: -q --score-min L,0,-0.2 --ignore-quals\n",
      "\n",
      "Now starting the Bowtie 2 aligner for CTreadCTgenome (reading in sequences from SRR6176401.filtered.fastq.gz_C_to_T.fastq with options -q --score-min L,0,-0.2 --ignore-quals --norc)\n",
      "Using Bowtie 2 index: /Users/bartosz/theta_methyl/genome/Bisulfite_Genome/CT_conversion/BS_CT\n",
      "\n",
      "Found first alignment:\tSRR6176401.1_1/1\t4\t*\t0\t0\t*\t*\t0\t0\tTTTTATATTTTTATTTTTTATAAATTATAATTTATTTTTATTGTATTTATGGTATTGGATTAATAATATTAGTGTAGTTAAATTAGTTAAAAGATTAATG\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIFIIIIIIIIFFFIIIIIIIIIFFIIIIIIFIIFF07BFIIIFFFFFFFFFFFFFFFB\tYT:Z:UU\n",
      "Now starting the Bowtie 2 aligner for CTreadGAgenome (reading in sequences from SRR6176401.filtered.fastq.gz_C_to_T.fastq with options -q --score-min L,0,-0.2 --ignore-quals --nofw)\n",
      "Using Bowtie 2 index: /Users/bartosz/theta_methyl/genome/Bisulfite_Genome/GA_conversion/BS_GA\n",
      "\n",
      "Found first alignment:\tSRR6176401.1_1/1\t4\t*\t0\t0\t*\t*\t0\t0\tTTTTATATTTTTATTTTTTATAAATTATAATTTATTTTTATTGTATTTATGGTATTGGATTAATAATATTAGTGTAGTTAAATTAGTTAAAAGATTAATG\tBBBFFFFFFFFFFIIIIIIIIIIIIIIIIIIIIIIIIFIIIIIFIIIIIIIIFFFIIIIIIIIIFFIIIIIIFIIFF07BFIIIFFFFFFFFFFFFFFFB\tYT:Z:UU\n",
      "\n",
      ">>> Writing bisulfite mapping results to SRR6176401.filtered_bismark_bt2.bam <<<\n",
      "\n",
      "sh: line 1: 86594 Abort trap: 6           /Users/bartosz/anaconda3/bin/samtools view -bSh - 2> /dev/null > SRR6176401.filtered_bismark_bt2.bam\n",
      "\n",
      "Reading in the sequence file SRR6176401.filtered.fastq.gz\n",
      "Processed 1000000 sequences so far\n",
      "Processed 2000000 sequences so far\n",
      "Processed 3000000 sequences so far\n",
      "Processed 4000000 sequences so far\n",
      "Processed 5000000 sequences so far\n",
      "Processed 6000000 sequences so far\n",
      "Processed 7000000 sequences so far\n",
      "Processed 8000000 sequences so far\n",
      "Processed 9000000 sequences so far\n",
      "Processed 10000000 sequences so far\n",
      "Processed 11000000 sequences so far\n",
      "Processed 12000000 sequences so far\n",
      "Chromosomal sequence could not be extracted for\tSRR6176401.12585343_12585343/1\t1\t30427571\n",
      "Processed 13000000 sequences so far\n",
      "Processed 14000000 sequences so far\n",
      "Processed 15000000 sequences so far\n",
      "Processed 16000000 sequences so far\n",
      "Processed 17000000 sequences so far\n",
      "Processed 18000000 sequences so far\n",
      "18811118 reads; of these:\n",
      "  18811118 (18811118100.00 reads; of these:%) were unpaired; of these:\n",
      "\n",
      "    16758290 (89.09%) aligned 0 times\n",
      "      188111181568935 ( (8.34%) aligned exactly 1 time100.00\n",
      "%    ) were unpaired; of these:483893\n",
      " (    2.5716730807% () aligned >1 times88.94\n",
      "%10.91) aligned 0 times%\n",
      " overall alignment rate    \n",
      "1544461 (8.21%) aligned exactly 1 time\n",
      "    535850 (2.85%) aligned >1 times\n",
      "11.06% overall alignment rate\n",
      "Processed 18811118 sequences in total\n",
      "\n",
      "Failed to close filehandle OUT: Broken pipe at /Users/bartosz/anaconda3/bin/bismark line 2436, <IN> line 150488944.\n",
      "Failed to close filehandle AMBIG: Bad file descriptor at /Users/bartosz/anaconda3/bin/bismark line 2437, <IN> line 150488944.\n",
      "Failed to close filehandle UNMAPPED: Bad file descriptor at /Users/bartosz/anaconda3/bin/bismark line 2438, <IN> line 150488944.\n",
      "\n",
      "Successfully deleted the temporary file SRR6176401.filtered.fastq.gz_C_to_T.fastq\n",
      "\n",
      "Final Alignment report\n",
      "======================\n",
      "Sequences analysed in total:\t18811118\n",
      "Number of alignments with a unique best hit from the different alignments:\t3095690\n",
      "Mapping efficiency:\t16.5%\n",
      "\n",
      "Sequences with no alignments under any condition:\t15122871\n",
      "Sequences did not map uniquely:\t592557\n",
      "Sequences which were discarded because genomic sequence could not be extracted:\t1\n",
      "\n",
      "Number of sequences with unique best (first) alignment came from the bowtie output:\n",
      "CT/CT:\t1530132\t((converted) top strand)\n",
      "CT/GA:\t1565557\t((converted) bottom strand)\n",
      "GA/CT:\t0\t(complementary to (converted) top strand)\n",
      "GA/GA:\t0\t(complementary to (converted) bottom strand)\n",
      "\n",
      "Number of alignments to (merely theoretical) complementary strands being rejected in total:\t0\n",
      "\n",
      "Final Cytosine Methylation Report\n",
      "=================================\n",
      "Total number of C's analysed:\t50952465\n",
      "\n",
      "Total methylated C's in CpG context:\t1849198\n",
      "Total methylated C's in CHG context:\t870005\n",
      "Total methylated C's in CHH context:\t1812390\n",
      "Total methylated C's in Unknown context:\t1273\n",
      "\n",
      "Total unmethylated C's in CpG context:\t4711566\n",
      "Total unmethylated C's in CHG context:\t6292601\n",
      "Total unmethylated C's in CHH context:\t35416705\n",
      "Total unmethylated C's in Unknown context:\t23743\n",
      "\n",
      "C methylated in CpG context:\t28.2%\n",
      "C methylated in CHG context:\t12.1%\n",
      "C methylated in CHH context:\t4.9%\n",
      "C methylated in Unknown context (CN or CHN):\t5.1%\n",
      "\n",
      "\n",
      "Bismark completed in 0d 0h 41m 27s\n",
      "\n",
      "====================\n",
      "Bismark run complete\n",
      "====================\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!bismark --bowtie2 genome/ SRR6176401.filtered.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r\n",
      "CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.\r\n",
      "To initialize your shell, run\r\n",
      "\r\n",
      "    $ conda init <SHELL_NAME>\r\n",
      "\r\n",
      "Currently supported shells are:\r\n",
      "  - bash\r\n",
      "  - fish\r\n",
      "  - tcsh\r\n",
      "  - xonsh\r\n",
      "  - zsh\r\n",
      "  - powershell\r\n",
      "\r\n",
      "See 'conda init --help' for more information and options.\r\n",
      "\r\n",
      "IMPORTANT: You may need to close and restart your shell after running 'conda init'.\r\n",
      "\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!conda activate py2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
